### ISSUE TITLE
VM Exception while processing transaction

### ISSUE THREAD

**Issue by mmsqe** [kinecosystem/kin-token#15]
Any insight of the below error when running `truffle test`?

`Contract: BasicToken approve, allowance, transferFrom should not allow approve() without resetting spender allowance to 0:
     AssertionError: Expected throw, got Error: VM Exception while processing transaction: revert instead`

**Comment by oryband** [comment]
Please make sure you are running the same truffle version as stated in the contracts.

**Comment by shoenseiwaso** [comment]
@mmsqe within the Kin repo top level directory, did you try:

```sh
$ npm install
$ npm test
```

Instead of `truffle test`?

**Comment by mmsqe** [comment]
Thanks for the replies, my version is:
Truffle v4.0.4 (core: 4.0.4)
Solidity v0.4.18 (solc-js)

while the issue I find might be that web3 doesn't support getBlockNumber when running truffle test, instead of getting timestamp directly `now = web3.eth.getBlock(web3.eth.blockNumber).timestamp;`
I need to do like:
```
function getTimestamp() {
  return new Promise((resolve, reject) => {
    web3.eth.getBlockNumber((err, number) => {
      web3.eth.getBlock(number, true, (err, block) => {
        resolve(block.timestamp);
      });
    });
  });
}
now = await getTimestamp();
```

another question is why there's no check for revert inside test/helpers/expectRevert.js like:
```
const revert = error.message.search('revert') >= 0;
assert(invalidOpcode || outOfGas || revert)
```

**Comment by shoenseiwaso** [comment]
> Thanks for the replies, my version is:
> Truffle v4.0.4 (core: 4.0.4)
> Solidity v0.4.18 (solc-js)

Per @oryband, you need to ensure you have the same version as stated in the contracts. Or more specifically, in the `package.json` file, which is:

```json
{

   "dependencies": {

      "truffle": "3.4.9",

   }
}
```

Truffle 3.4.9 itself includes a much lower version of solc-js (I can't recall which one) than v0.4.18. Can you try re-running with:

```sh
$ npm install
$ npm test
```

And see if that works?

**Comment by mmsqe** [comment]
`npm test` works fine for the timestamp issue, since if I run truffle test, that's using my global truffle (4.0.4), while node run with the truffle@3.4.9 from node_module
however, I still need to add revert in the assert to make `npm test` pass

**Comment by mmsqe** [comment]
Steps to reproduce:
1) run `ganache-cli`
2) run `npm test`

Current result:
4 tests pass with 1 test fails
![2018-01-10 11 24 58](https://user-images.githubusercontent.com/2867936/34780231-81599160-f65d-11e7-8a32-cde2fd5badb1.png)

Expected result:
all tests pass
<img width="255" alt="screen shot 2018-01-10 at 11 25 36 pm" src="https://user-images.githubusercontent.com/2867936/34780285-b11e430a-f65d-11e7-91ba-e52a43aef05c.png">

p.s.
I can pass all the tests when connecting with your rpc without any change, 
but I need to add revert in the assert inside test/helpers/expectRevert.js when connecting to my existing testrpc instance (running `ganache-cli`)

**Comment by shoenseiwaso** [comment]
Sounds like a difference between `ganache-cli` and `testrpc`. First thought is gas limit was hit? I'm not familiar with `ganache-cli`, but do you have it configured with `$accounts` as setup in `scripts/common.sh` with the first two accounts unlocked (see `scripts/test.sh`)?

**Comment by shoenseiwaso** [comment]
Just googled the [documentation](https://github.com/trufflesuite/ganache-cli#using-ganache-cli), but it looks like `ganache-cli` supports the same command line options as `test-rpc` (critically, `-u` and `--account`) so maybe just try replacing `testrpc` with `ganache-cli` in `scripts/test.sh` and then try running `npm test` again?

**Comment by mmsqe** [comment]
Sorry for the late reply, it might be related to this [issue](https://github.com/trufflesuite/ganache-cli/issues/435). When executes [revert()](https://github.com/kinfoundation/kin-token/blob/master/contracts/BasicToken.sol#L24), expectRevert.js gets error msg as 'invalid opcode' when running with testrpc, but 'revert' when running with ganache-cli.

Test under Ganache CLI v6.0.3 (ganache-core: 2.0.2)

**Comment by oryband** [comment]
Hey. Try editing `expectRevert.js` in the following manner:
```javascript
// expectRevert.js

// Change this:
const invalidOpcode = error.message.search('invalid opcode') > -1;

// To this:
const invalidOpcode = error.message.search('revert') >= -1 || error.message.search('invalid opcode') > -1;
```
Lemme know how it works out.

**Comment by mmsqe** [comment]
I guess you mean `search('revert') >= 0`,  this works both when I run `truffle test` and `npm test`, I will create a PR later.

**Comment by oryband** [comment]
fixed in #16
